### PR TITLE
[MenuBundle] Some refactoring of the menuBundle - Make it extendable

### PR DESCRIPTION
**Pull request by denbatte**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets |

### CODE REVIEW

File: src/Kunstmaan/MenuBundle/Repository/MenuItemRepository.php [hunk 1]
```diff
@@ -2,9 +2,30 @@
 
 namespace Kunstmaan\MenuBundle\Repository;
 
+use Doctrine\ORM\EntityManager;
 use Gedmo\Tree\Entity\Repository\NestedTreeRepository;
 
 class MenuItemRepository extends NestedTreeRepository
 {
+    public function getMenuItemsForLanguage(EntityManager $em, $menuName, $locale)
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Repository/MenuItemRepository.php]: What is the EntityManager doing here as an argument? It should be removed as this is a repository, which already has access to the entity manager.

File: src/Kunstmaan/MenuBundle/Resources/config/services.yml [hunk 2]
```diff
@@ -2,6 +2,8 @@ parameters:
     kunstmaan_menu.menu.adaptor.class: Kunstmaan\MenuBundle\Service\MenuAdaptor
     kunstmaan_menu.menu.service.class: Kunstmaan\MenuBundle\Service\MenuService
     kunstmaan_menu.menu.twig.extension.class: Kunstmaan\MenuBundle\Twig\MenuTwigExtension
+    kunstmaan_menu.menuitem.entity.class: 'KunstmaanMenuBundle:MenuItem'
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Resources/config/services.yml]: Be consistent : use the class name including namespace.

File: src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php [hunk 3]
```diff
@@ -18,14 +18,18 @@ class MenuTwigExtension extends \Twig_Extension
...
      * @param EntityManager $em
      * @param RouterInterface $router
+     * @param $repositoryClass
      */
-    public function __construct(EntityManager $em, RouterInterface $router)
+    public function __construct(EntityManager $em, RouterInterface $router, $repositoryClass)
     {
         $this->em = $em;
         $this->router = $router;
+        $this->repositoryClass = $repositoryClass;
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php]: You might want to define an interface for the repository class, so the class can be checked and an exception can be thrown if it does not implement the interface (which should contain the definition of the getMenuItemsForLanguage method).

File: src/Kunstmaan/MenuBundle/Entity/BaseMenuItem.php [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+            return $this->getTitle();
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getSelectTitle()
+    {
+        if ($this->getType() == self::TYPE_PAGE_LINK) {
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Entity/BaseMenuItem.php]: Why was this copy/pasted, it's exactly the same as getDisplayTitle so you should call that here instead...

File: src/Kunstmaan/MenuBundle/Entity/Menu.php [hunk 5]
```diff
@@ -2,123 +2,21 @@
...
-        $item->setMenu($this);
-
-        $this->items->add($item);
-    }
-
-    /**
-     * @param MenuItem $item
-     */
-    public function removeItem(MenuItem $item)
+    public function getAdminListConfiguratorClass()
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Entity/Menu.php]: Why is this here (it is not used anywhere)? Or is something missing?

File: src/Kunstmaan/MenuBundle/Entity/MenuItem.php [hunk 6]
```diff
@@ -15,410 +12,23 @@
...
-        }
-
-        return false;
-    }
-
-    /**
-     * @param ExecutionContextInterface $context
-     */
-    public function validateEntity(ExecutionContextInterface $context)
+    public function getAdminListConfiguratorClass()
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Entity/MenuItem.php]: Are you sure this PR is complete? This method is not called anywhere.

File: src/Kunstmaan/MenuBundle/Entity/MenuItem.php [hunk 7]
```diff
@@ -15,410 +12,23 @@
...
-
-        return $this;
-    }
-
     /**
+     * Get the classname of the menuItem formtype.
      * @return string
      */
-    public function getTitle()
+    public function getFormTypeClass()
```
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Entity/MenuItem.php]: Are you sure this PR is complete? This method is not called anywhere.

File: src/Kunstmaan/MenuBundle/Entity/Menu.php (hunk 5 above)
Comment (wimvds) [inline src/Kunstmaan/MenuBundle/Entity/Menu.php]: Are you sure this PR is complete? This method is not called anywhere.

**Comment by jockri** [comment]
The MenuItem class and Menu class should be defined in the bundle parameters. Because they are not only used in the twig extension but also in the adminlist controllers etc.

File: src/Kunstmaan/MenuBundle/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -26,6 +26,11 @@ public function getConfigTreeBuilder()
                     ->defaultValue(array())
                     ->prototype('scalar')->end()
                 ->end()
+                ->scalarNode('menu_entity')->defaultValue('Kunstmaan\MenuBundle\Entity\Menu')->end()
+                ->scalarNode('menuitem_entity')->defaultValue('Kunstmaan\MenuBundle\Entity\MenuItem')->end()
+                ->scalarNode('menu_adminlist')->defaultValue('Kunstmaan\MenuBundle\AdminList\MenuAdminListConfigurator')->end()
+                ->scalarNode('menuitem_adminlist')->defaultValue('Kunstmaan\MenuBundle\AdminList\MenuItemAdminListConfigurator')->end()
+                ->scalarNode('menuitem_form')->defaultValue('Kunstmaan\MenuBundle\Form\MenuItemAdminType')->end()
             ->end();
```
Comment (jockri) [inline src/Kunstmaan/MenuBundle/DependencyInjection/Configuration.php]: You should also have menu_form when you have menu_entity

**Comment by jockri** [comment]
In the MenuItemAdminListController, there are still some references to `KunstmaanMenuBundle:MenuItem`this should be replaced.

File: src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php [hunk 9]
```diff
@@ -50,44 +51,30 @@ public function getFunctions()
...
         $activeClass = false;
-
         if (isset($options['activeClass'])) {
             $activeClass = $options['activeClass'];
         }
 
         $options = array_merge($this->getDefaultOptions($activeClass), $options);
-        $html = $repo->buildTree($arrayResult, $options);
+
+	$html = $menuRepo->buildTree($arrayResult, $options);
```
Comment (n8green) [inline src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php]: I think $menuRepo either needs to be defined before this point, or you need to use $this->repository instead

Comment (aistis-) [inline src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php]: :+1:

File: src/Kunstmaan/MenuBundle/Entity/BaseMenu.php [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+     * @ORM\OneToMany(targetEntity="Kunstmaan\MenuBundle\Entity\BaseMenuItem", mappedBy="menu", cascade={"persist", "remove"}, orphanRemoval=true)
+     */
+    protected $items;
+
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+	$this->items = new ArrayCollection();
```
Comment (aistis-) [inline src/Kunstmaan/MenuBundle/Entity/BaseMenu.php]: Please configure your IDE for spacing :/

File: src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php [hunk 11]
```diff
@@ -50,44 +51,30 @@ public function getFunctions()
...
```
Comment (aistis-) [inline src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php]: $menuRepo is not defined!

File: src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php [hunk 12]
```diff
@@ -50,44 +51,30 @@ public function getFunctions()
...
-            ->andWhere('m.name = :name')
-            ->setParameter('name', $name)
-            ->andWhere('(nt.online = 1 OR nt.online IS NULL)')
-            ->andWhere('(n.deleted = 0 OR n.deleted IS NULL)')
-            ->getQuery();
-        $arrayResult = $query->getArrayResult();
-
-        // Make sure the parent item is not offline
-        $foundIds = array();
+	/** @var MenuItem $menuRepo */
```
Comment (aistis-) [inline src/Kunstmaan/MenuBundle/Twig/MenuTwigExtension.php]: What's that mate?
